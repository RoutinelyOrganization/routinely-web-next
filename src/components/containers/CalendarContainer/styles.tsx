'use client';

import media from '@/styles/mediaQueries';
import { colors, fonts } from '@/styles/variables';
import styled from 'styled-components';

interface IChangeDisplayCalendar {
  $openCalendar: boolean;
}

export const Container = styled.section`
  position: relative;
  max-width: 100%;
  width: 100%;
  display: flex;
  justify-content: space-between;
  align-items: center;

  > h1 {
    font-size: ${fonts.sizes.xxlarge};
  }
  @media screen and (max-width: 890px) {
    > h1 {
      font-size: ${fonts.sizes.large};
      margin-bottom: 16px;
    }
  }

  ${media.tablet} {
    > h1 {
      font-size: ${fonts.sizes.medium};
      margin-bottom: 0;
    }
  }

  ${media.mobile} {
    display: none;
  }
`;

export const ContainerIcons = styled.div`
  position: relative;
  display: flex;
  align-items: center;
  max-width: 20%;
  width: 100%;
  justify-content: space-around;
  gap: 8px;

  > img {
    display: inline-block;
    background-color: ${colors.lightgray};
    border-radius: 8px;
    width: 50px;
    &:hover {
      cursor: pointer;
    }
  }

  ${media.desktop} {
    position: absolute;
    flex-direction: column;
    right: 0;
    top: 0;

    > img {
      margin-bottom: 27px;
    }
  }

  ${media.tablet} {
    > img {
      width: 40px;
      height: 32px;
    }
  }
`;

export const ContainerCalendar = styled.div<IChangeDisplayCalendar>`
  position: absolute;
  top: 100%;
  left: -30%;
  z-index: 10;
  display: ${({ $openCalendar }) => ($openCalendar ? 'block' : 'none')};

  ${media.desktop} {
    left: -145px;
    top: 40%;
  }

  ${media.tablet} {
    left: -180px;
  }
`;

export const SecondContainerCalendar = styled.div`
  display: none;
  width: 100%;

  ${media.mobile} {
    display: block;
  }
`;
