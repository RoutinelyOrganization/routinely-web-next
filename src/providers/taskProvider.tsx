'use client';

import { typeTaskOptions } from '@/constants/typeTask';
import type { ActionFormExecuteType, ExecuteServiceType } from '@/contexts/TaskContext';
import { TaskContext } from '@/contexts/TaskContext';
import { makeCreateTask } from '@/factories/services/makeCreateTask';
import { makeDeleteTask } from '@/factories/services/makeDeleteTask';
import { makeUpdateTask } from '@/factories/services/makeUpdateTask';
import type { Task } from '@/types/task';
import type { TypeTask } from '@/types/typeTasks';
import { useEffect, useState } from 'react';

interface ITaskProvider {
  children: React.ReactNode;
}

export const TaskProvider: React.FC<ITaskProvider> = ({ children }) => {
  const [tasks, setTasks] = useState<Task[]>([]);
  const [selectedTask, setSelectedTask] = useState<Task | null>(null);
  const [formIsOpen, setFormIsOpen] = useState<boolean>(false);
  const [selectedTypeTaskCtx, setSelectedTypeTask] = useState<TypeTask['type'] | null>(null);
  const [selectedTypeTask, setSelectedTypeTaskCtx] = useState<TypeTask | null>(null);
  const [selectedActionForm, setActionForm] = useState<ActionFormExecuteType>({
    openConfirm: false,
    action: null,
  });
  const [executeServiceTask, setExecuteServiceTask] = useState<ExecuteServiceType>({
    execute: async () => {},
  });

  useEffect(() => {
    switch (selectedActionForm.action) {
      case 'create':
        setExecuteServiceTask({
          execute: async (data?: any) => {
            const { ok, body } = await makeCreateTask(selectedTask!, data.token);

            if (!ok) return ok;

            setTasks([...tasks, { ...selectedTask!, id: body.id }]);
            setSelectedTask(null);

            return ok;
          },
        });
        break;
      case 'update':
        setExecuteServiceTask({
          execute: async (data?: any) => {
            const { ok } = await makeUpdateTask(selectedTask! || data.task, data.token);

            if (!ok) return ok;

            const id = selectedTask?.id || data.task.id;
            const updatedTask = selectedTask || data.task;
            setTasks(tasks.map(task => (task.id === id ? updatedTask : task)));
            setSelectedTask(null);

            return ok;
          },
        });
        break;
      case 'delete':
        setExecuteServiceTask({
          execute: async (data?: any) => {
            await makeDeleteTask(selectedTask!.id, data.token);

            setTasks(tasks.filter(task => task.id !== selectedTask!.id));
            setSelectedTask(null);
            return true;
          },
        });
        break;
    }

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [selectedTask, selectedActionForm]);

  useEffect(() => {
    const typeTask = typeTaskOptions.filter(item => item.type === selectedTypeTaskCtx)[0];
    setSelectedTypeTaskCtx(typeTask);
  }, [selectedTypeTaskCtx]);
  return (
    <TaskContext.Provider
      value={{
        tasks,
        setTasks,
        selectedTask,
        setSelectedTask,
        formIsOpen,
        setFormIsOpen,
        selectedTypeTask,
        setSelectedTypeTask,
        setActionForm,
        executeServiceTask,
        selectedActionForm,
      }}
    >
      {children}
    </TaskContext.Provider>
  );
};
