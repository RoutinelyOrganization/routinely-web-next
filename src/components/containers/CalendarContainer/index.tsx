'use client';

import DateCalendar from '@/components/calendar';
import ContainerPrevNext from '@/components/containers/ContainerPrevNext';
import { useCalendar } from '@/hooks/useCalendar';
import iconCalendar from '@public/icons/Frame 56.svg';
import Image from 'next/image';
import { useEffect, useState } from 'react';
import * as S from './styles';
export default function CalendarContainer() {
  const [openCalendar, setOpenCalendar] = useState(false);
  const [changeDate, setChangeDate] = useState<'prev' | 'next' | null>(null);
  const { day, nameMonth, weekDay, year, setDate, date } = useCalendar();

  useEffect(() => {
    if (changeDate === 'next') {
      setDate(date.add(1, 'day'));
      setChangeDate(null);
      return;
    }
    if (changeDate === 'prev') {
      setDate(date.subtract(1, 'day'));
      setChangeDate(null);
      return;
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [changeDate]);

  const onChangeDate = (value: 'prev' | 'next') => {
    setChangeDate(value);
  };

  return (
    <>
      <S.Container>
        {weekDay && day && nameMonth && year && (
          <h1>
            {weekDay}, {day} de {nameMonth} de {year}
          </h1>
        )}
        <S.ContainerIcons>
          <Image
            src={iconCalendar}
            alt="CalendÃ¡rio"
            onClick={() => setOpenCalendar(!openCalendar)}
          />
          <ContainerPrevNext onChange={onChangeDate} />
          <S.ContainerCalendar $openCalendar={openCalendar}>
            <DateCalendar />
          </S.ContainerCalendar>
        </S.ContainerIcons>
      </S.Container>
      <S.SecondContainerCalendar>
        <DateCalendar />
      </S.SecondContainerCalendar>
    </>
  );
}
