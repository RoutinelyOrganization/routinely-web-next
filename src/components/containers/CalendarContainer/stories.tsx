import { CalendarProvider } from '@/providers/calendarProvider';
import { GlobalStyles } from '@/styles/globalStyles';
import CalendadrContainer from '.';

export default {
  title: 'containers/CalendaderContainer',
  component: CalendadrContainer,
};

export const Template = () => {
  return (
    <div style={{ maxWidth: '1024px' }}>
      <CalendarProvider>
        <GlobalStyles />
        <CalendadrContainer />
      </CalendarProvider>
    </div>
  );
};
