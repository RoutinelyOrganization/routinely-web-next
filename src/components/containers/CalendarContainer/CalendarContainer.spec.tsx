import { CalendarProvider } from '@/providers/calendarProvider';
import { fireEvent, render, screen } from '@testing-library/react';
import dayjs from 'dayjs';
import React from 'react';
import CalendarContainer from '.';

const today = dayjs();

const weekDay = (index: number): string => {
  const weekDays = [
    'Domingo',
    'Segunda-feira',
    'Terça-feira',
    'Quarta-feira',
    'Quinta-feira',
    'Sexta-feira',
    'Sábado',
  ];
  return weekDays[index];
};

const monthName = (index: number): string => {
  const monthNames = [
    'Janeiro',
    'Fevereiro',
    'Março',
    'Abril',
    'Maio',
    'Junho',
    'Julho',
    'Agosto',
    'Setembro',
    'Outubro',
    'Novembro',
    'Dezembro',
  ];
  return monthNames[index];
};

describe('<CalendarContainer/>', () => {
  it('should render', () => {
    render(
      <CalendarProvider>
        <CalendarContainer />
      </CalendarProvider>,
    );
    const heading = screen.getByRole('heading');
    expect(heading.textContent).toBe(
      `${weekDay(today.day())}, ${today.date()} de ${monthName(today.month())} de ${today.year()}`,
    );

    const [iconCalendar, iconPrev, iconNext] = screen.getAllByRole('img');
    expect(iconCalendar).toHaveAttribute('alt', 'Calendário');
    expect(iconPrev).toHaveAttribute('alt', 'Icone para voltar');
    expect(iconNext).toHaveAttribute('alt', 'Icone para avançar');
  });

  it('should change title to prev date on clin in icon prev', () => {
    render(
      <CalendarProvider>
        <CalendarContainer />
      </CalendarProvider>,
    );
    const heading = screen.getByRole('heading');
    expect(heading.textContent).toBe(
      `${weekDay(today.day())}, ${today.date()} de ${monthName(today.month())} de ${today.year()}`,
    );
    const prevIcon = screen.getByRole('img', { name: 'Icone para voltar' });

    // should change week day and date

    const prevDay = today.subtract(1, 'day');
    fireEvent.click(prevIcon);
    expect(heading.textContent).toBe(
      `${weekDay(prevDay.day())}, ${prevDay.date()} de ${monthName(prevDay.month())} de ${today.year()}`,
    );

    // should change week day, date and month
    for (let i = 1; i < today.date(); i++) {
      fireEvent.click(prevIcon);
    }
    const prevMonth = today.subtract(today.date(), 'day');

    expect(heading.textContent).toBe(
      `${weekDay(prevMonth.day())}, ${prevMonth.date()} de ${monthName(prevMonth.month())} de ${today.year()}`,
    );
  });

  it('should change title to next date on clin in icon next', () => {
    render(
      <CalendarProvider>
        <CalendarContainer />
      </CalendarProvider>,
    );
    const heading = screen.getByRole('heading');
    expect(heading.textContent).toBe(
      `${weekDay(today.day())}, ${today.date()} de ${monthName(today.month())} de ${today.year()}`,
    );
    const nextIcon = screen.getByRole('img', { name: 'Icone para avançar' });

    //should change week day and date
    const nextDay = today.add(1, 'day');
    fireEvent.click(nextIcon);
    expect(heading.textContent).toBe(
      `${weekDay(nextDay.day())}, ${nextDay.date()} de ${monthName(nextDay.month())} de ${today.year()}`,
    );

    // should change week day, date and month
    const maxNumberOfClick = 31 - today.date();
    for (let i = 1; i < maxNumberOfClick; i++) {
      fireEvent.click(nextIcon);
    }
    const nextMonth = today.add(maxNumberOfClick, 'day');
    expect(heading.textContent).toBe(
      `${weekDay(nextMonth.day())}, ${nextMonth.date()} de ${monthName(nextMonth.month())} de ${today.year()}`,
    );
  });

  it('should change visible calendar', async () => {
    const setOpenCalendar = jest.fn();
    jest.spyOn(React, 'useState').mockImplementation((init?: boolean) => [init, setOpenCalendar]);

    render(
      <CalendarProvider>
        <CalendarContainer />
      </CalendarProvider>,
    );

    const icon = screen.getByAltText('Calendário');
    fireEvent.click(icon);

    expect(setOpenCalendar).toHaveBeenCalledWith(true);
  });
});
