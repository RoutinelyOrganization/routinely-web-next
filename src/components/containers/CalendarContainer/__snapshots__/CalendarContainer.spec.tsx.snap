// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CalendarContainer/> should match snapshot 1`] = `
.c2 {
  max-width: 140px;
  width: 100%;
  background-color: #ebeaed;
  display: flex;
  align-items: center;
  justify-content: space-around;
  border-radius: 4px;
}

.c2 >span {
  width: 2px;
  height: 16px;
  background-color: #d8d6dc;
}

.c2 >img {
  width: 40px;
  padding: 8px;
}

.c2 >img:hover {
  cursor: pointer;
}

.c0 {
  position: relative;
  max-width: 100%;
  width: 100%;
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.c0 >h1 {
  font-size: 2.5rem;
}

.c1 {
  position: relative;
  display: flex;
  align-items: center;
  max-width: 20%;
  width: 100%;
  justify-content: space-around;
  gap: 8px;
}

.c1 >img {
  display: inline-block;
  background-color: #ebeaed;
  border-radius: 2px;
  width: 50px;
}

.c1 >img:hover {
  cursor: pointer;
}

.c3 {
  position: absolute;
  top: 100%;
  left: -30%;
  z-index: 10;
  display: none;
}

.c4 {
  display: none;
  width: 100%;
}

@media (max-width: 904px) {
  .c2 {
    max-width: 200px;
  }

  .c2 >img {
    width: 36px;
    padding: 5px;
  }
}

@media (max-width: 680px) {
  .c2 {
    max-width: 200px;
  }

  .c2 >img {
    width: 36px;
    padding: 5px;
  }
}

@media screen and (max-width:890px) {
  .c0 >h1 {
    font-size: 2rem;
    margin-bottom: 16px;
  }
}

@media (max-width: 680px) {
  .c0 >h1 {
    font-size: 1.5rem;
    margin-bottom: 0;
  }
}

@media (max-width: 500px) {
  .c0 {
    display: none;
  }
}

@media (max-width: 904px) {
  .c1 {
    position: absolute;
    flex-direction: column;
    right: 0;
    top: 0;
  }

  .c1 >img {
    margin-bottom: 27px;
  }
}

@media (max-width: 680px) {
  .c1 >img {
    width: 40px;
    height: 32px;
  }
}

@media (max-width: 904px) {
  .c3 {
    left: -145px;
    top: 40%;
  }
}

@media (max-width: 680px) {
  .c3 {
    left: -180px;
  }
}

@media (max-width: 500px) {
  .c4 {
    display: block;
  }
}

<div>
  <section
    class="c0"
  >
    <h1>
      Quarta-feira
      , 
      14
       de 
      Agosto
       de 
      2024
    </h1>
    <div
      class="c1"
    >
      <img
        alt="Calendário"
        src="[object Object]"
      />
      <div
        class="c2"
      >
        <img
          alt="Icone para voltar"
          src="[object Object]"
        />
        <span />
        <img
          alt="Icone para avançar"
          src="[object Object]"
        />
      </div>
      <div
        class="c3"
      >
        <div
          data-testid="mock-calendario"
        />
      </div>
    </div>
  </section>
  <div
    class="c4"
  >
    <div
      data-testid="mock-calendario"
    />
  </div>
</div>
`;
